### DISCUSSION TITLE
v2.0 Release Candidate - Migration Guide

### DISCUSSION THREAD

**Discussion by endigo9740** [skeletonlabs/skeleton#1854] [General]
### [*** Skeleton v2 is now live! ***](https://github.com/skeletonlabs/skeleton/discussions/1947)

---

# Migration Guide

If you would like like participate in testing the Skeleton v2 release candidate, please follow the instructions provided below.

> [!important]
> Unlike standard Skeleton releases, the v2 release candidate is designated as a `pre-release`, meaning there's a few extra steps to move from `v1.x` to `v2 RC`. Please read and follow the instruction below very carefully.

# Using the Skeleton CLI

To install and test a new Skeleton v2 RC project using the Skelton CLI, use the `@rc` tag instead `@latest`

```
npm create skeleton-app@rc my-skeleton-app
    - Enable Typescript when prompted (recommended)
cd my-skeleton-app
```

# Migrating v1.x Projects

1. Uninstall the Skeleton v1.x package.

```
npm uninstall @skeletonlabs/skeleton
```

2. Install the Skeleton v2 RC package and the new standalone Skeleton Tailwind plugin.

```
npm add -D @skeletonlabs/skeleton@rc @skeletonlabs/tw-plugin@rc
```

3. If your SvelteKit projects uses Typescript, install the standard Node type definitions:

```
npm add -D @types/node
```

4. Add the following Tailwind directives to the top of your global stylesheet `/src/app.postcss`

```
@tailwind base;
@tailwind components;
@tailwind utilities;
@tailwind variants;

/* (all other styles here...) */
```

5. Remove the Skeleton stylesheet imports in `/src/routes/+layout.svelte`

```diff
<script>
	// Remove:
-	import '@skeletonlabs/skeleton/themes/theme-skeleton.css';
-	import '@skeletonlabs/skeleton/styles/skeleton.css';

	import '../app.postcss'; // KEEP THIS!
</script>
```

> NOTE: this includes any preset or custom generated themes. We'll cover theme migration in detail below.

6. (OPTIONAL) Install [vite-plugin-tailwind-purgecss](https://skeleton-docs-git-v2-skeleton-labs.vercel.app/docs/purgecss) to dramatically reduce your production CSS bundle size.

# Migrating your Tailwind Configuration

Next, we'll need to migrate `tailwind.config.cjs` to either a Typescript or Javascript format. If your project has Typescript enabled, use `.ts` (Typescript) otherwise please opt for `.js` (Javascript).

<details>
	<summary><h3>Typescript (tailwind.config.ts)</h3></summary>
	
```ts
import { join } from 'path';
import type { Config } from 'tailwindcss';

// 1. Import the Skeleton plugin
import { skeleton } from '@skeletonlabs/tw-plugin';

const config = {
	// 2. Opt for dark mode to be handled via the class method
	darkMode: 'class',
	content: [
		'./src/**/*.{html,js,svelte,ts}',
		// 3. Append the path to the Skeleton package
		join(require.resolve(
			'@skeletonlabs/skeleton'),
			'../**/*.{html,js,svelte,ts}'
		)
	],
	theme: {
		extend: {},
	},
	plugins: [
		// 4. Append the Skeleton plugin (after other plugins)
		skeleton
	]
} satisfies Config;

export default config;
```
</details>

<details>
	<summary><h3>Javascript (tailwind.config.js)</h3></summary>

```js
// @ts-check
import { join } from 'path';

// 1. Import the Skeleton plugin
import { skeleton } from '@skeletonlabs/tw-plugin';

/** @type {import('tailwindcss').Config} */
export default {
	// 2. Opt for dark mode to be handled via the class method
	darkMode: 'class',
	content: [
		'./src/**/*.{html,js,svelte,ts}',
		// 3. Append the path to the Skeleton package
		join(require.resolve(
			'@skeletonlabs/skeleton'),
			'../**/*.{html,js,svelte,ts}'
		)
	],
	theme: {
		extend: {},
	},
	plugins: [
		// 4. Append the Skeleton plugin (after other plugins)
		skeleton
	]
}
```
</details>

# Migrating Themes

Theme configuration is now handled via your Skeleton Tailwind plugin settings in `tailwind.config`.

> NOTE: You can mix `preset` and `custom` themes if you wish!

<details>
	<summary><h3>Preset Themes</h3></summary>
	
1. Register one or more themes within the plugin settings:

```ts
plugins: [
	skeleton({
		themes: {
			// Register each theme within this array:
			preset: [ "skeleton", "modern", "crimson" ] 
		}
	})
]
```

2. Open `/src/app.html` and set your active theme via the `data-theme` attribute on the `body` element:

```html
<body data-theme="skeleton">
```

</details>

<details>
	<summary><h3>Custom Themes</h3></summary>
	
Skeleton themes now use a `CSS-in-JS` format. If you wish to migrate your v1.x theme, use one of the following techniques:

## Option 1) Use the Theme Generator

Recreate your theme using the updated [Theme Generator](https://skeleton-docs-git-v2-skeleton-labs.vercel.app/docs/generator), which provides themes in the new format.

## Option 2) Migrate to the v2 CSS-in-JS format

1. Visit https://transform.tools/css-to-js
2. Copy the contents within your theme's `:root` selector and paste them into the CSS field on the left
3. Copy the converted property values on the right (NOTE: the data within the `converted` object, not `converted` itself).

![img](https://i.imgur.com/17VTP7m.gif)

4. Create a new file in the root of your project, such as `my-custom-theme.[ts|js]` to house your new theme.

5. Copy the following template into this file, then add your converted CSS-in-JS theme properties to the `properties` field:

> NOTE: you may also use `.js` and strip out the `CustomThemeConfig` Typescript type definition

```ts
import type { CustomThemeConfig } from '@skeletonlabs/tw-plugin';

export const myCustomTheme: CustomThemeConfig = {
    name: 'my-custom-theme',
    properties: {
		// (PASTE YOUR THEME PROPERTIES HERE)
	}
}
```

6. Import your new theme in `tailwind.config` and update your Skeleton plugin settings as follows:

```ts
import { myCustomTheme } from './my-custom-theme';

// ...

plugins: [
	skeleton({
		themes: {
			custom: [ myCustomTheme ]
		}
	})
]
```

7. Open `/src/app.html` and set your active theme via the `data-theme` attribute on the `body` element:

```html
<body data-theme="my-custom-theme">
```

</details>

# Migrating Core Features

The following features include breaking changes that may affect your project.

<details>
	<summary><h3>Table of Contents</h3></summary>
	
This features has been reimplemented from the ground up. We recommend you [follow the updated documentation](https://skeleton-docs-git-v2-skeleton-labs.vercel.app/utilities/table-of-contents).
</details>

<details>
	<summary><h3>Drawer, Modal, and Toast Stores</h3></summary>
	
We've refactored the stores for all singleton-based overlay features to help prevent known [SvelteKit SSR issues](https://github.com/skeletonlabs/skeleton/wiki/SvelteKit-SSR-Warning). This includes two notable changes:

1. You must now initialize the stores via `initializeStore()` in your root layout.

```ts
import { initializeStores } from '@skeletonlabs/skeleton';

initializeStores();
```

2. You now import and access the stores in a new manner.

```ts
import { getDrawerStore } from "@skeletonlabs/skeleton";

const drawerStore = getDrawerStore();
```

View the documentation for each respective feature for more information:

- Drawers: https://skeleton-docs-git-v2-skeleton-labs.vercel.app/utilities/drawers
- Modals: https://skeleton-docs-git-v2-skeleton-labs.vercel.app/utilities/modals
- Toasts: https://skeleton-docs-git-v2-skeleton-labs.vercel.app/utilities/toasts
</details>

<details>
	<summary><h3>Skeleton Data Tables</h3></summary>
	
This feature has been completely removed in Skeleton v2. We now recommend [Svelte Simple Datatables](https://github.com/vincjo/datatables) paired with [Skeleton's Table Element](https://www.skeleton.dev/elements/tables) styles.
</details>

<details>
	<summary><h3>Deprecated Autocomplete Props</h3></summary>
	
The deprecated `whitelist | blacklist` props have been removed, please move to `allowlist | denylist` respectively 
</details>

<details>
	<summary><h3>Legacy Typography System</h3></summary>
	
The legacy "on-by-default" Typography system has now been completely phased out. Please migrate to the new opt-in system covered in the [Typography documentation](https://skeleton-docs-git-v2-skeleton-labs.vercel.app/elements/typography). In most cases this requires appending a single class to each typographic element.
</details>

<details>
	<summary><h3>Paginator Offset Prop</h3></summary>
	
The `PaginatorSettings` parameter called `offset` has now been renamed to `page` for better semantics.
</details>

**Comment by dev-guy** [comment]
Regarding "Legacy Typography System", I was already using class="h1" etc.. In 1.x, fonts are thick. In 2.x, fonts are thin. I'm using the skeleton theme. Maybe there's something I'm not aware of.

**Reply by endigo9740** [reply]
@dev-guy you're referring to what are now called theme "enhancements":
https://skeleton-docs-git-v2-skeleton-labs.vercel.app/docs/themes#enhancements

Some themes have a set of extra styles provided, and you can enable this using this feature. This includes some adjustment to font headings for certain themes. Enable this feature and things should display as they did prior.

**Comment by alexvdvalk** [comment]
Remember to include:
```typescript
plugins: [
    require('@tailwindcss/forms'),
    // ...
  ],
  ```
  into the tailwind config if using tailwind forms.

**Reply by endigo9740** [reply]
Thanks for the suggestion! FYI I'll be revising the migration guide for the v2 stable release next week. I'll aim to include this, as it's tripped up a few people in testing.
